File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    
    dic = {}
    for idx, value in enumerate(players):
        dic[value] = idx

    for c in callings:
        pre , idx = dic[c] - 1 , dic[c]
        players[pre], players[idx] = players[idx] , players[pre]
        dic[players[pre]] , dic[players[idx]] = dic[players[pre]] - 1 , dic[players[idx]] + 1

    return players

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.HashMap;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        HashMap<String, Integer> playerIndexMap = new HashMap<>();
        for (int i = 0; i < players.length; i++) {
            playerIndexMap.put(players[i], i);
        }

        for (String calling : callings) {
            int currentIndex = playerIndexMap.get(calling);
            if (currentIndex > 0) { 
                String tmp = players[currentIndex - 1];

                players[currentIndex] = tmp;
                players[currentIndex - 1] = calling;
                
                playerIndexMap.put(tmp, currentIndex);
                playerIndexMap.put(calling, currentIndex - 1);
            }
        }

        return players;
    }
}
